File: typescript/examples/05_human_in_the_loop.ts
/**
 * Parmana TypeScript SDK
 *
 * Example 05
 *
 * Human-in-the-Loop
 *
 * Demonstrates an authorized human override
 * becoming part of the immutable Execution
 * Trust Record.
 */

import {
  Authority,
  Authorization,
  BusinessTransaction,
  Execution,
  ExecutionTrustRecord,
  Intent,
  Override,
  PolicyReference,
} from "../src/index.js";

async function main(): Promise<void> {
  const authority: Authority = {
    authorityId: "authority-001",
    authorityName: "Acme Robotics",
    createdAt: new Date(),
  };

  const authorization: Authorization = {
    authorizationId: "authorization-001",
    authorityId: authority.authorityId,
    subject: "warehouse-robot-01",
    permissions: [
      "MOVE_PALLET",
    ],
    issuedAt: new Date(),
    expiresAt: new Date(
      Date.now() + 24 * 60 * 60 * 1000,
    ),
  };

  const intent: Intent = {
    intentId: "intent-001",
    authorizationId: authorization.authorizationId,
    operation: "MOVE_PALLET",
    target: "Restricted Zone",
    createdAt: new Date(),
  };

  const policy: PolicyReference = {
    policyName: "warehouse-policy",
    policyVersion: "1.0.0",
  };

  const transaction: BusinessTransaction = {
    businessTransactionId: "txn-001",
    authority,
    authorization,
    intent,
    policy,
    createdAt: new Date(),
  };

  const override: Override = {
    overrideId: "override-001",
    businessTransactionId:
      transaction.businessTransactionId,
    approvedBy: "operations-manager",
    reason:
      "Urgent customer shipment requires manual approval.",
    justification:
      "Customer escalation approved by warehouse supervisor.",
    approvedAt: new Date(),
  };

  const trustRecord = {
    trustRecordId: "trust-record-001",

    businessTransactionId:
      transaction.businessTransactionId,

    transaction,

    overrides: [
      override,
    ],

    executions: [] as Execution[],

    verifications: [],

    receipts: [],

    trustRecordHash:
      "3df9ef8d2cb14dc78eab1259d2fa4c91",

    createdAt: new Date(),

    updatedAt: new Date(),
  } satisfies ExecutionTrustRecord;

  console.log();
  console.log("======================================");
  console.log(" Parmana TypeScript SDK");
  console.log(" Example 05 - Human in the Loop");
  console.log("======================================");
  console.log();

  console.log("Business Transaction");
  console.log("------------------------------");
  console.log(
    "Transaction :",
    transaction.businessTransactionId,
  );
  console.log(
    "Operation   :",
    intent.operation,
  );
  console.log(
    "Target      :",
    intent.target,
  );

  console.log();

  console.log("Policy");
  console.log("------------------------------");
  console.log(
    "Policy Name :",
    policy.policyName,
  );
  console.log(
    "Version     :",
    policy.policyVersion,
  );

  console.log();

  console.log("Human Override");
  console.log("------------------------------");
  console.log(
    "Override ID :",
    override.overrideId,
  );
  console.log(
    "Approved By :",
    override.approvedBy,
  );
  console.log(
    "Reason      :",
    override.reason,
  );
  console.log(
    "Approved At :",
    override.approvedAt,
  );

  console.log();

  console.log("Execution Trust Record");
  console.log("------------------------------");
  console.log(
    "Trust Record :",
    trustRecord.trustRecordId,
  );
  console.log(
    "Overrides    :",
    trustRecord.overrides.length,
  );
  console.log(
    "Hash         :",
    trustRecord.trustRecordHash,
  );

  console.log();

  console.log("Trust Chain");
  console.log("------------------------------");
  console.log("✓ Authority");
  console.log("✓ Authorization");
  console.log("✓ Intent");
  console.log("✓ Policy Reference");
  console.log("✓ Human Override");
  console.log("✓ Execution Trust Record");

  console.log();

  console.log(
    "The override becomes an immutable part",
  );
  console.log(
    "of the Execution Trust Record and is",
  );
  console.log(
    "available for replay, verification,",
  );
  console.log(
    "and independent audit.",
  );

  console.log();
  console.log("Example completed.");
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: python/examples/05_human_in_the_loop.py
"""
Parmana Python SDK

Example 05

Human in the Loop

Demonstrates an authorized human override after a
policy rejects an execution request.

Authority
        ↓
Authorization
        ↓
Intent
        ↓
Policy Evaluation
        ↓
REJECTED
        ↓
Human Override
        ↓
Approved
        ↓
Execution
"""

from datetime import datetime
from uuid import uuid4

from parmana.models.authority import Authority
from parmana.models.authorization import Authorization
from parmana.models.business_transaction import BusinessTransaction
from parmana.models.decision import (
    Decision,
    DecisionOutcome,
)
from parmana.models.execution import (
    Execution,
    ExecutionMode,
    ExecutionStatus,
)
from parmana.models.execution_trust_record import (
    ExecutionTrustRecord,
)
from parmana.models.intent import Intent
from parmana.models.override import Override
from parmana.models.policy_reference import PolicyReference


def main() -> None:

    authority = Authority(
        authority_id=str(uuid4()),
        authority_type="ORGANIZATION",
        authority_name="Acme Corporation",
        created_at=datetime.utcnow(),
        public_key="public-key",
        signature_algorithm="Ed25519",
    )

    authorization = Authorization(
        authorization_id=str(uuid4()),
        authority_id=authority.authority_id,
        subject="warehouse-robot-01",
        action="ENTER",
        resource="restricted-zone",
        issued_at=datetime.utcnow(),
    )

    intent = Intent(
        intent_id=str(uuid4()),
        operation="ENTER_RESTRICTED_ZONE",
        target="warehouse-zone-b",
        parameters={
            "reason": "Emergency maintenance",
        },
    )

    policy = PolicyReference(
        policy_name="warehouse-safety-policy",
        policy_version="1.0.0",
    )

    transaction = BusinessTransaction(
        business_transaction_id=str(uuid4()),
        authority=authority,
        authorization=authorization,
        intent=intent,
        policy_reference=policy,
        created_at=datetime.utcnow(),
    )

    #
    # Initial policy decision.
    #
    decision = Decision(
        decision_id=str(uuid4()),
        intent_id=intent.intent_id,
        policy=policy,
        signals={
            "restricted_zone": True,
            "emergency": False,
        },
        outcome=DecisionOutcome.REJECTED,
        reason="Restricted zone access denied.",
        evaluated_at=datetime.utcnow(),
    )

    print()
    print("========================================")
    print("Initial Policy Decision")
    print("========================================")

    print(f"Decision : {decision.outcome.value}")
    print(f"Reason   : {decision.reason}")

    #
    # Human override.
    #
    override = Override(
        override_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        approved_by="Operations Manager",
        reason="Emergency maintenance approved.",
        justification="Critical repair required.",
        approved_at=datetime.utcnow(),
    )

    print()
    print("Human Override Approved")
    print("-------------------------------")

    print(f"Approved By : {override.approved_by}")
    print(f"Reason      : {override.reason}")

    #
    # Execution after override.
    #
    execution = Execution(
        execution_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        decision=decision,
        status=ExecutionStatus.COMPLETED,
        mode=ExecutionMode.SYNC,
        started_at=datetime.utcnow(),
        completed_at=datetime.utcnow(),
        evidence={
            "override": True,
            "operator": override.approved_by,
            "result": "Access Granted",
        },
    )

    record = ExecutionTrustRecord(
        trust_record_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        transaction=transaction,
        overrides=(override,),
        executions=(execution,),
        verifications=(),
        receipts=(),
        trust_record_hash="trust-record-hash",
        created_at=datetime.utcnow(),
        updated_at=datetime.utcnow(),
    )

    print()
    print("========================================")
    print("Execution Trust Record")
    print("========================================")

    print(f"Transaction : {record.business_transaction_id}")
    print(f"Overrides   : {len(record.overrides)}")
    print(f"Executions  : {len(record.executions)}")

    print()

    print("Human-in-the-loop execution completed.")

    print("========================================")


if __name__ == "__main__":
    main()